// script/index.js
const vm = require('vm')

function Script (x, y, size, method, body) {
	this.position = {x, y}
	this.size = size
	this.method = method
	this.body = body
	this.compile()
}

Script.prototype.setSandbox = function (sandbox) {
	this.sandbox = sandbox
}

Script.prototype.addToStage = function(stage) {
	if (!this.sprite) {
		this.sprite = Resources.getSprite('script.png')
		this.sprite.position.x = this.position.x * this.size
		this.sprite.position.y = this.position.y * this.size
		this.sprite.width = this.size
		this.sprite.height = this.size
	}

	this.stage = stage
	this.stage.addChild(this.sprite)
}

Script.prototype.removeFromStage = function() {
	if (this.stage && this.sprite) {
		this.stage.removeChild(this.sprite)
	}
}

Script.prototype.compile = function () {
	this.script = new vm.Script(this.body)
}

Script.prototype.run = function () {
	if (!this.sandbox) this.sandbox = {}
	const sandbox = Object.create(this.sandbox)
	const context = new vm.createContext(sandbox)
	this.script.runInContext(context)
}

module.exports = Script
