// input/index.js
let keys = []

const update = function () {
	for (let i = 0; i < keys.length; i++) {
		if (keys[i].phase === 'down') {
			keys[i].phase = 'pressed'
		}
		else if (keys[i].phase === 'up') {
			keys.splice(i, 1)
		}
	}
}

document.onkeydown = function (event) {
	if (keys.indexOf(event.code) === -1) {
		let key = {
			code: event.code,
			phase: 'down'
		}
		keys.push(key)
	}
}

document.onkeyup = function (event) {
	for (let i = 0; i < keys.length; i++) {
		if (keys[i].code === event.code) {
			keys[i].phase = 'up'
		}
	}
}

const isKeyPressed = function (keyCode) {
	for (let i = 0; i < keys.length; i++) {
		if (keys[i].code === keyCode) {
			return keys[i].phase === 'pressed'
		}
	}
	return false
}

const isKeyDown = function (keyCode) {
	for (let i = 0; i < keys.length; i++) {
		if (keys[i].code === keyCode) {
			return keys[i].phase === 'down'
		}
	}
	return false
}

const isKeyUp = function (keyCode) {
	for (let i = 0; i < keys.length; i++) {
		if (keys[i].code === keyCode) {
			return keys[i].phase === 'up'
		}
	}
	return false
}

const onmousemove = function (callback) {
	document.onmousemove = function (event) {
		callback(event)
	}
}

const onclick = function (callback) {
	document.onclick = function (event) {
		callback(event)
	}
}

const onwheel = function (callback) {
	document.onwheel = function (event) {
		callback(event)
		event.preventDefault()
	}
}

module.exports = {
	update: update,
	isKeyPressed: isKeyPressed,
	isKeyDown: isKeyDown,
	onmousemove: onmousemove,
	onclick: onclick,
	onwheel: onwheel
}
