// world/index.js
function World (layers, width, height) {
	this.name = "Rosewood"
	this.width = width
	this.height = height
	this.tileSize = 32
	this.layers = layers

	this.tiles = []
	this.scripts = []
	this.regions = []
}

World.prototype.createScript = function (x, y, method, body) {
	let script = new Script(x, y, this.tileSize, method, body)
	this.scripts.push(script)
	script.addToStage(scriptLayer)
	script.sprite.tint = '0xFF0000'
	return script
}

World.prototype.getScript = function (x, y) {
	for (var i = 0; i < this.scripts.length; i++) {
		let script = this.scripts[i]
		if (script.position.x === x && script.position.y === y) {
			return script
		}
	}
}

World.prototype.removeScript = function (x, y) {
	let script = this.getScript(x, y)
	if (script) {
		script.removeFromStage()
		let index = this.scripts.indexOf(script)
		this.scripts.splice(index, 1)
		DebugInfo.update()
	}
}

World.prototype.createTile = function (x, y, texture, options) {
	if (x >= 0 && x < this.width && y >= 0 && y < this.height) {

		let layer = 0

		if (options) {
			if (options.layer !== undefined) {
				layer = options.layer
			}
		}

		this.removeTile(x, y, layer)

		let tile = new Tile(x, y, this.tileSize, texture, options)
		let stage = this.layers[layer]
		tile.addToStage(stage)

		this.tiles.push(tile)

		DebugInfo.count = this.tiles.length
		DebugInfo.update()
	}
}

World.prototype.getTile = function (x, y, layer) {
	for (let i = 0; i < this.tiles.length; i++) {
		let tile = this.tiles[i]
		if (layer !== undefined && tile.options !== undefined && tile.options.layer !== undefined) {
			if (tile.getPosition().x === x && tile.getPosition().y === y && tile.options.layer === layer) {
				return tile
			}
		} else if (tile.getPosition().x === x && tile.getPosition().y === y) {
			return tile
		}
	}
}

World.prototype.removeTile = function (x, y, layer) {
	let tile = this.getTile(x, y, layer)
	if (tile) {
		tile.removeFromStage()
		let index = this.tiles.indexOf(tile)
		this.tiles.splice(index, 1)
		DebugInfo.count = this.tiles.length
		DebugInfo.update()
	}
}

World.prototype.wipe = function() {
	this.tiles = []
	for (let i = 0; i < this.layers.length; i++) {
		this.layers[i].children = []
	}
}

World.prototype.load = function(file, callback) {
	let path = __dirname + '/maps/' + file + '.json'
	let data = fs.readFileSync(path).toString()
	let map = JSON.parse(data)

	this.wipe()

	for (let texture in map.tiles) {
		for (let i = 0; i < map.tiles[texture].length; i++) {
			let tileObj = map.tiles[texture][i]
			this.createTile(tileObj.x, tileObj.y, texture, tileObj.options)
		}
	}

	for (var i = 0; i < map.scripts.length; i++) {
		let scriptObj = map.scripts[i]
		this.createScript(scriptObj.x, scriptObj.y, scriptObj.method, scriptObj.body)
	}

	if (callback) {
		callback(map.options)
	}
}

World.prototype.save = function(file, options, callback) {
	let map = {
		name: this.name,
		width: this.width,
		height: this.height
	}

	if (options) {
		map.options = options
	}

	let tilesObj = {}

	for (var i = 0; i < this.tiles.length; i++) {
		let tile = this.tiles[i]

		let tileObj = {
			x: tile.getPosition().x,
			y: tile.getPosition().y,
			options: tile.options
		}

		if (tilesObj[tile.texture] === undefined) {
			tilesObj[tile.texture] = []
		}

		tilesObj[tile.texture].push(tileObj)
	}

	map.tiles = tilesObj

	let scriptsArr = []

	for (var i = 0; i < this.scripts.length; i++) {
		let script = this.scripts[i]

		let scriptObj = {
			x: script.position.x,
			y: script.position.y,
			method: script.method,
			body: script.body,
		}

		scriptsArr.push(scriptObj)
	}

	map.scripts = scriptsArr

	let data = JSON.stringify(map, null, '\t')

	let path = __dirname + '/maps/' + file + '.json'
	fs.writeFile(path, data, function () {
		if (callback) {
			callback()
		}
	})
}

module.exports = World
