// alert-view/index.js
const csjs = require('csjs-inject')

const css = csjs`
	.alert {
		width: 400px;
		height: 150px;
		background: #FFF;
		color: #333;
		font-family: sans-serif;
		padding: 20px;
		border-radius: 10px;
		box-shadow: 0 0 30px black;
		position: fixed;
		left: 50%;
		top: 50%;
		margin-left: -200px;
		margin-top: -200px;
		display: none;
	}
`

let alertElement
let canvas
let buttons = []
let initialized = false
let open = false

function close() {
	canvas.style.webkitFilter = 'none'
	alertElement.style.display = 'none'
	open = false
}

function init() {
	canvas = document.querySelector('canvas')
	alertElement = document.createElement('div')
	alertElement.setAttribute('class', css.alert)
	document.body.appendChild(alertElement)
	initialized = true
}

function show(message) {
	if (initialized) {
		canvas.style.webkitFilter = 'blur(2px)'
		alertElement.innerHTML = message
		alertElement.style.display = 'block'
		open = true
	} else {
		console.log('alert-view has not been initialized')
	}
}

module.exports = {
	init: init,
	show: show,
	close: close,
	isOpen: () => open
}
