// editor/index.js
let editor_instance = undefined

let editor_containerElement = undefined
let editor_methodElement = undefined
let editor_saveElement = undefined
let editor_closeElement = undefined
let editor_removeElement = undefined

let editor_callback = undefined

let _editor_isOpen = false

function init(editor_containerId, editor_methodId, editor_codeId, editor_saveId, editor_closeId, editor_removeId) {
	editor_instance = ace.edit(editor_codeId)
	editor_instance.setTheme("ace/theme/twilight")
	editor_instance.session.setMode("ace/mode/javascript")

	editor_containerElement	= document.querySelector('#' + editor_containerId)
	editor_methodElement	= document.querySelector('#' + editor_methodId)
	editor_saveElement		= document.querySelector('#' + editor_saveId)
	editor_closeElement		= document.querySelector('#' + editor_closeId)
	editor_removeElement 	= document.querySelector('#' + editor_removeId)

	if (editor_containerElement.style.display === 'block') {
		_editor_isOpen = true
	} else {
		_editor_isOpen = false
	}

	editor_saveElement.addEventListener('click', function () {
		editor_close('save')
	})

	editor_closeElement.addEventListener('click', function () {
		editor_close('cancel')
	})

	editor_removeElement.addEventListener('click', function () {
		editor_close('remove')
	})
}

function editor_close(state) {
	editor_containerElement.style.display = 'none'
	if (editor_callback) {
		let method = editor_methodElement.value
		let script = editor_instance.getValue()
		editor_callback(state, method, script)
	}
}

function editor_open(method, script, callback) {
	editor_methodElement.value = method
	editor_instance.setValue(script)
	editor_containerElement.style.display = 'block'
	editor_callback = callback
}

function editor_isOpen() {
	return _editor_isOpen
}

function editor_getCode() {
	return editor_instance.getValue()
}

function editor_setCode(code) {
	return editor_instance.setValue(code)
}

module.exports = {
	init: init,
	close: editor_close,
	open: editor_open,
	isOpen: editor_isOpen,
	setCode: editor_setCode,
	getCode: editor_getCode
}
