// debug-info/index.js
const DebugInfo = {
	mode: 'play',
	layer: 0,
	walkable: true,
	tile: 'stone',
	count: 0,
	modeText:		new PIXI.Text('Mode: play',		{fill: 'white', fontSize: 14}),
	layerText:		new PIXI.Text('Layer: 1',		{fill: 'white', fontSize: 14}),
	walkableText:	new PIXI.Text('Walkable: true',	{fill: 'white', fontSize: 14}),
	tileText:		new PIXI.Text('Tile: stone',	{fill: 'white', fontSize: 14}),
	countText:		new PIXI.Text('Count: 0',		{fill: 'white', fontSize: 14}),
	init: function (stage, width, height) {
		this.stage = stage
		this.stage.addChild(this.modeText)
		this.stage.addChild(this.layerText)
		this.stage.addChild(this.walkableText)
		this.stage.addChild(this.tileText)
		this.stage.addChild(this.countText)
		this.update()
		this.relocate(width, height)
	},
	update: function () {
		this.modeText.text		= 'Mode: '		+ this.mode
		this.countText.text 	= 'Count: '		+ this.count
		this.walkableText.text 	= 'Walkable: '	+ this.walkable
		this.layerText.text 	= 'Layer: '		+ (this.layer + 1)
		this.tileText.text		= 'Tile: '		+ this.tile

		this.modeText.visible		= (this.mode === 'play' 	? false : true)
		this.countText.visible		= (this.mode === 'play' 	? false : true)
		this.layerText.visible		= (this.mode === 'play'		? false : true)
		this.walkableText.visible	= (this.mode === 'paint' 	? true : false)
		this.tileText.visible		= (this.mode === 'paint' 	? true : false)
	},
	relocate: function (width, height) {
		this.modeText.position.x = width - 160
		this.modeText.position.y = 20

		this.countText.position.x = width - 160
		this.countText.position.y = 40
		
		this.layerText.position.x = width - 160
		this.layerText.position.y = 60
		
		this.walkableText.position.x = width - 160
		this.walkableText.position.y = 80

		this.tileText.position.x = width - 160
		this.tileText.position.y = 100
	}
}

module.exports = DebugInfo
